_posts/2022-06-18-zio-streams.md

```diff
@@ -0,0 +1,537 @@
+recognize the scope of context you are working in - i.e Are we processing the
+content of a file, which is finite, or are we streaming generated events to a
+websocket, which could be infinite.
+
+## Example: Processing Files
+
+Let's walk through a simple, but working example application that parses
+markdown blog entries, automatically looks for tagged words, and re-generates a
+file with tags automatically added, and linked to their respective page. We'll
+also generate a "search" file, that is a JSON object that could be used by a
+JavaScript front end to find all pages that use a particular tag.
+
+### The Content
+
+We'd normally read these from files, but for the purpose of this post, we'll
+have the example markdown as a String containing the sample content.
+
+```scala
+  val post1: String = "hello-word.md"
+  val post1_content: Array[Byte] =
+    """---
+      |title: "Hello World"
+      |tags: []
+      |---
+      |======
+      |
+      |## Generic Heading
+      |
+      |Even pretend blog posts need a #generic intro.
+      |""".stripMargin.getBytes
+
+  val post2: String = "scala-3-extensions.md"
+  val post2_content: Array[Byte] =
+    """---
+      |title: "Scala 3 for You and Me"
+      |tags: []
+      |---
+      |======
+      |
+      |## Cool Heading
+      |
+      |This is a post about #Scala and their re-work of #implicits via thing like #extensions.
+      |""".stripMargin.getBytes
+
+  val post3: String = "zio-streams.md"
+  val post3_content: Array[Byte] =
+    """---
+      |title: "ZIO Streams: An Introduction"
+      |tags: []
+      |---
+      |======
+      |
+      |## Some Heading
+      |
+      |This is a post about #Scala and #ZIO #ZStreams!
+""".stripMargin.getBytes
+
+  val fileMap: Map[String, Array[Byte]] = Map(
+    post1 -> post1_content,
+    post2 -> post2_content,
+    post3 -> post3_content
+  )
+```
+
+We'll take note that we have `tags: []` in the post front-matter, and our
+content has words prefixed with a `#` that we want to index on. We'll also make
+a helper `Map` to represent how we'd call the file name to get the content
+later, as if we were reading from actual files.
+
+### Pipelines
+
+Let's break up our concerns, and cover a first step of collecting all the tags
+we'll need, and then use those results for the blog post regeneration. Finally,
+we'll put everything together into an application we can run start-to-end.
+
+#### Collecting tags
+
+Let's outline the process of collecting our tags. The general steps we want to
+take are:
+
+1. Filter words that match our tagging pattern
+2. Remove any punctuation, in case the last work in the sentence is tag - as
+   well as to remove the `#`.
+3. Convert our parsed tags to lowercase
+4. Put them all in a `Set[String]` to avoid duplicates.
+
+Let's also set up some helpers to looks for our `#tag`, as well as a reusable
+regex to remove punctuation when we need to.
+
```

Comment:
I really like the progress of the example up to this point. Reading this, it's clear what is expected and how to get there.